src/components/codemirror.tsx

```diff
@@ -45,22 +58,156 @@ export default function CodeMirror(props: CodeMirrorProps) {
 		});
 	};
 
-
-	useEffect(() => {
+	useSignalEffect(() => {
+		if(!isNewSaveStrat.value) return
+		if(yProviderAwarenessSignal.value === undefined) return;
+		yProviderAwarenessSignal.value.on("change", () => {
+			yProviderAwarenessSignal.value?.getStates().forEach((state) => {
+				console.log(state)
+				try{
+					if(props.persistenceState === undefined) return;
+					if(state.saved == "saved"){
+						let persistenceState = props.persistenceState.peek();
+						if(persistenceState.kind === "PERSISTED" && persistenceState.game !== "LOADING"){
+							props.persistenceState.value = {...persistenceState, cloudSaveState: "SAVED"};
+						}
+					} else if(state.saved == "error"){
+						let persistenceState = props.persistenceState.peek();
+						if(persistenceState.kind === "PERSISTED" && persistenceState.game !== "LOADING"){
+							props.persistenceState.value = {...persistenceState, cloudSaveState: "ERROR"};
+						}
+					}
+				} catch(e){
+					console.log(e)
+				}	
+			});
+		});
+	});
+	useSignalEffect(() => {
+		if(editorRef !== undefined) {
+			editorRef.destroy()
+		}
 		if (!parent.current) throw new Error('Oh golly! The editor parent ref is null')
 
-		const editor = new EditorView({
-			state: createEditorState(props.initialCode ? props.initialCode : '', () => {
-				if (editor.state.doc.toString() === lastCode) return
-				lastCode = editor.state.doc.toString()
-				onCodeChangeRef.current?.()
-			}, () => onRunShortcutRef.current?.()),
-			parent: parent.current,
-		})
-
-		setEditorRef(editor);
-		props.onEditorView?.(editor)
-	}, [])
+		if(!isNewSaveStrat.value || (props.roomState?.value.roomId === "" || props.persistenceState?.peek().session === null)){
+			const editor = new EditorView({
+				state: createEditorState(props.initialCode ? props.initialCode : '', () => {
+					if (editor.state.doc.toString() === lastCode) return
+					lastCode = editor.state.doc.toString()
+					onCodeChangeRef.current?.()
+				}, () => onRunShortcutRef.current?.()),
+				parent: parent.current,
+			})
+
+			setEditorRef(editor);
+			props.onEditorView?.(editor)
+			return
+		}
+
+		if(!props.roomState) return
+		if(!props.persistenceState) return
+		try{
+			if(yDoc !== undefined){
+				yDoc.destroy();
+			}
+			if(provider !== undefined){
+				provider.destroy();
+			}
+			yDoc = new Y.Doc();
+			provider = new WebrtcProvider(props.roomState.value.roomId, yDoc, {
+				signaling: [
+					"wss://yjs-signaling-server-5fb6d64b3314.herokuapp.com",
```

Comment:
this should not be a string literal, but instead in some sort of environment var